File: crates/taflow-core/src/indicators/volume_by_time_profile.rs
use crate::error::{TaError, TaResult};

/// Borrowed average-volume values for every configured intraday time bucket.
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct VolumeByTimeProfileValue<'a> {
    pub bins: &'a [f64],
}

/// Running arithmetic mean of volume grouped by time of day.
#[derive(Debug, Clone)]
pub struct VolumeByTimeProfile {
    buckets: usize,
    offset: i32,
    sums: Vec<f64>,
    counts: Vec<usize>,
    bins: Vec<f64>,
    available: bool,
}

impl VolumeByTimeProfile {
    /// Create an empty profile with evenly sized daily buckets.
    pub fn new(buckets: usize, utc_offset_minutes: i32) -> TaResult<Self> {
        if buckets == 0 {
            return Err(TaError::InvalidParameter {
                name: "buckets",
                value: buckets.to_string(),
                reason: "must be positive",
            });
        }
        Ok(Self {
            buckets,
            offset: utc_offset_minutes,
            sums: vec![0.0; buckets],
            counts: vec![0; buckets],
            bins: vec![0.0; buckets],
            available: false,
        })
    }

    fn bucket(&self, timestamp: i64) -> usize {
        const DAY_NS: i128 = 86_400_000_000_000;
        let local = timestamp as i128 + self.offset as i128 * 60_000_000_000;
        (local.rem_euclid(DAY_NS) * self.buckets as i128 / DAY_NS) as usize
    }

    /// Append one chronological OHLCV bar and Unix-nanosecond timestamp.
    pub fn append(
        &mut self,
        _open: f64,
        _high: f64,
        _low: f64,
        _close: f64,
        volume: f64,
        timestamp: i64,
    ) -> Option<VolumeByTimeProfileValue<'_>> {
        let bucket = self.bucket(timestamp);
        self.sums[bucket] += volume;
        self.counts[bucket] += 1;
        self.bins[bucket] = self.sums[bucket] / self.counts[bucket] as f64;
        self.available = true;
        self.value()
    }

    /// Return all current bucket means after at least one bar exists.
    pub fn value(&self) -> Option<VolumeByTimeProfileValue<'_>> {
        self.available
            .then_some(VolumeByTimeProfileValue { bins: &self.bins })
    }

    /// Restore the profile to its newly constructed state without reallocating.
    pub fn reset(&mut self) {
        self.sums.fill(0.0);
        self.counts.fill(0);
        self.bins.fill(0.0);
        self.available = false;
    }
}

File: python/taflow/indicators/volume_by_time_profile.py
"""Volume-by-time profile adapter."""

from typing import Any

import numpy as np

from .._native import VolumeByTimeProfile as _Native
from .._series import as_float64_series, as_int64_series


class VolumeByTimeProfile:
    """Compute expanding mean volume by intraday time bucket.

    Rust assigns each bar's volume to an equal-width local-day bucket and owns
    all persistent sums, counts, and history. Output shape is
    ``(bars, buckets)`` and is available from the first bar. This class maps to
    Wickra ``VolumeByTimeProfile``.

    Args:
        open: Required opening-price history.
        high: Required high-price history.
        low: Required low-price history.
        close: Required closing-price history.
        volume: Required volume history used by the profile.
        timestamp: Chronological Unix timestamps in nanoseconds.
        buckets: Number of equal-width local-day buckets. Defaults to 24.
        utc_offset_minutes: Fixed offset applied before bucketing. Defaults to 0.

    Raises:
        ValueError: If inputs differ in length or ``buckets`` is zero.
    """

    def __init__(self, buckets: int = 24, utc_offset_minutes: int = 0) -> None:
        """Initialize an empty configured native state.
        """
        self._state = _Native(buckets, utc_offset_minutes)

    def append(
        self,
        open: float,
        high: float,
        low: float,
        close: float,
        volume: float,
        timestamp: int,
    ) -> "VolumeByTimeProfile":
        """Append one OHLCV bar and Unix-nanosecond timestamp; return ``self``."""
        self._state.append(
            float(open), float(high), float(low), float(close),
            float(volume), int(timestamp),
        )
        return self

    def extend(
        self,
        open: Any,
        high: Any,
        low: Any,
        close: Any,
        volume: Any,
        timestamp: Any,
    ) -> "VolumeByTimeProfile":
        """Append aligned OHLCV and timestamp histories and return ``self``."""
        arrays = tuple(
            as_float64_series(item) for item in (open, high, low, close, volume)
        ) + (as_int64_series(timestamp),)
        if len({len(item) for item in arrays}) != 1:
            raise ValueError("OHLCV and timestamp inputs must have equal lengths")
        self._state.extend(*arrays)
        return self

    @property
    def value(self) -> list[float] | None:
        """Return current bucket volume means, or ``None`` before any bar."""
        return self._state.value

    def compute(self) -> np.ndarray:
        """Return aligned ``(bars, buckets)`` native output history."""
        return self._state.compute()

    def reset(self) -> "VolumeByTimeProfile":
        """Clear native state and history and return ``self``."""
        self._state.reset()
        return self

    def __len__(self) -> int:
        """Return the processed-bar count delegated to native state."""
        return len(self._state)
